### PR TITLE
support non-clean shutdown

### PR DESCRIPTION
**Pull request by tsahee**
Adds shutdown markers, and adds time on blockchain which invokes triedb cache writes

### CODE REVIEW

**Review by joshuacolvin0 — APPROVED** [review]
LGTM

File: core/blockchain.go [hunk 1]
```diff
@@ -1339,6 +1339,16 @@ func (bc *BlockChain) WriteBlockAndSetHead(block *types.Block, receipts []*types
...
 }
 
+// WriteBlockAndSetHeadWithTime also counts processTime, which will cause intermittent TrieDirty cache writes
+func (bc *BlockChain) WriteBlockAndSetHeadWithTime(block *types.Block, receipts []*types.Receipt, logs []*types.Log, state *state.StateDB, emitHeadEvent bool, processTime time.Duration) (status WriteStatus, err error) {
+	if !bc.chainmu.TryLock() {
+		return NonStatTy, errChainStopped
+	}
+	defer bc.chainmu.Unlock()
+	bc.gcproc += processTime
+	return bc.writeBlockAndSetHead(block, receipts, logs, state, emitHeadEvent)
```
Comment (PlasmaPower) [inline core/blockchain.go:1349]: Let's put this in blockchain_arbitrum.go to prevent merge conflicts

**Review by PlasmaPower — DISMISSED** [review]
LGTM

File: core/blockchain_arbitrum.go [hunk 2]
```diff
@@ -54,3 +67,52 @@ func (bc *BlockChain) ClipToPostNitroGenesis(blockNum rpc.BlockNumber) (rpc.Bloc
 	}
 	return blockNum, currentBlock
 }
+
+// finds the number of blocks that aren't prunable
+func (bc *BlockChain) FindRetentionBound() uint64 {
```
Comment (hkalodner) [inline core/blockchain_arbitrum.go:72]: If `TrieRetention` is 0, the behavior should be the same as standard geth, right? Might be good to have an early exit to make that clear

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: core/blockchain_arbitrum.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tsahee) [reply core/blockchain_arbitrum.go:72]: this would work better with a little-different search algo.. leaving it to later

**Review by hkalodner — APPROVED** [review]
LGTM

**Pull request merged**